src/main/kotlin/org/elm/lang/core/types/RecordFieldReferenceTable.kt

```diff
@@ -27,17 +43,30 @@ data class RecordFieldReferenceTable(
 
     /** Create a new table with references from this table and [other] */
     operator fun plus(other: RecordFieldReferenceTable): RecordFieldReferenceTable {
-        return copy(frozen = false).apply { addAll(other) }
+        val initialCapacity = refsByField.size + other.refsByField.keys.count { it !in refsByField }
+        val newRefs = HashMap<String, MutableSet<ElmNamedElement>>(initialCapacity, 1f)
+        refsByField.mapValuesTo(newRefs) { (field, set) ->
+            val otherSet = other.refsByField[field].orEmpty()
+            val initialSetCapacity = set.size + otherSet.count { it !in set }
+            newSet(initialSetCapacity).apply { otherSet.forEach { add(it) } }
+        }
+        other.refsByField.forEach { (k, s) ->
+            newRefs.getOrPut(k) { newSet(s.size).apply { s.forEach { add(it) } } }
+        }
+        return RecordFieldReferenceTable(newRefs)
     }
 
-    fun copy(frozen: Boolean = this.frozen): RecordFieldReferenceTable {
-        return RecordFieldReferenceTable(
-                refsByField.mapValuesTo(mutableMapOf()) { (_, v) -> v.toMutableSet() },
-                frozen
-        )
+    /** Prevent further modifications to this table. Tables cannot be unfrozen. */
+    fun freeze() {
```

Comment:
Reference tables are now frozen in place, avoiding copies just to freeze them. This mostly only helps in the type expression inference, since the freeze was done in the same pass as replacement for value expressions. Even in that case, this can reduce copies if no other replacements need to be made to part of a ty.